### ISSUE TITLE
Heist loads underscored templates with top level bindings

### ISSUE THREAD

**Issue by kaol** [snapframework/heist#74]
There's a problem with how compiled Heist loads templates in compileTemplates'. It processes all the templates, including ones starting with an underscore, even though those aren't meant to be used directly. This causes problems when namespaces are also used, since subtemplates potentially have tags that are only bound with withSplices.

https://github.com/kaol/heist/tree/nsbindcall has the fix and https://github.com/kaol/heist/tree/nsbindcall-test has a test for calling underscored template with namespaces. The new test should work with master and my branch but the code in master fails with the added underscored template in the nsbinderr test.

**Comment by mightybyte** [comment]
Hi there.  Thank you so much for your outstanding treatment of this problem.  Your separate branches for the test and fix make it very easy to investigate.  But I'm having trouble reproducing your results.  First of all, your test branch doesn't build for me.  You have the following code on line 188 of Heist.Compiled.Tests:

    res <- runEitherT $ do
        hs <- initHeist $ nsBindTemplateHC

But that has compile errors.  I think it should be this:

    res <- runExceptT $ do
        hs <- ExceptT $ initHeist $ nsBindTemplateHC

Could this have happened because you were building against the 0.14-stable branch at one point and then switched to the master branch?  That problem is easy for me to fix, but when I build your test branch with the above ExceptT fix, I get "namespace bind error test" when I would expect to get "namespace bind call test".  So I'm not sure what's going on her.  It looks like your nsbindcall-test branch was branched off of master, which is fine.  But just keep in mind that master is on 1.0 and since 1.0 hasn't been released to hackage yet, your work will have to be back ported (possibly just a simple cherry-pick) to the 0.14-stable branch to release it to hackage right now.

**Comment by kaol** [comment]
Right, I got 1.0 and 0.14 branches mixed up. Sorry about that. I pushed a fresh version at https://github.com/kaol/heist/tree/nsbindcall-test2 though it'll most likely be just the same changes you yourself did.

That's expected that nsbinderr gives out the error without the nsbindcall fix. Out of the three nsBind tests, nsBindErrorTest function is the only one that uses hcErrorNotBound .~ True. I wrote nsbindcall test to verify that _call.tpl is still usable with callTemplate even though compileTemplates' skips it. That part worked before and I wanted to check that it still does.

**Comment by mightybyte** [comment]
After thinking about this a bit, I'm not sure if this is the right way to
fix this issue.  The treatment of templates that begin with an underscore
is not a feature of heist itself.  It is a feature of the heist snaplet.
Starting a template name with an underscore really just says that you don't
want it to be routed by the snaplet's heistServe function, not that you
don't want it to be used directly at all.

So I'm wondering if maybe we should consider some other way of suppressing
the errors.  Note that I'm not categorically opposed to your solution.  I
just think that we should consider more options first.  One possibility
might be to make a splice that allows you to suppress errors as needed.
Another might be to let the user specify a list of templates that should
have errors suppressed.  Maybe there are even more that I haven't thought
of.  What do you think?

**Comment by kaol** [comment]
I expanded my approach a bit. I added a new parameter to SpliceConfig, compiledTemplateFilter, of type TPath -> Bool. This time I can't give a separate test branch of it since it's a new field. The code is available at https://github.com/kaol/heist/tree/compiledTemplateFilter and https://github.com/kaol/heist/tree/0.14-compiledTemplateFilter (backported).

Also, snap would need a change to use it: https://github.com/kaol/snap/tree/0.14-compiledTemplateFilter (I didn't make a 1.0 version).

The other option for this that I can think of, which would tie in with my hope for better error reporting would be to change the type of initHeist'. `EitherT [String] IO (HeistState n)` is, I think, too strict about its errors.  I'd rather have a function `initHeist'' :: ... -> IO (HeistState n)` and add a new field to HeistState, `_compileErrors :: [(TPath, SomeErrorType)]`, where SomeErrorType would have instance Show. The caller of initHeist'' could decide whether the compileErrors it got are anything it should care about.

**Comment by mightybyte** [comment]
I like what you've done in the compiledTemplateFilter branch.  I think I'd be willing to accept a pull request for that.  Don't won't worry about the 0.14 back port unless you really need it because we're going to have to release 1.0 soon due to GHC 8.
